File: dclib/core/src/main/java/dclib/graphics/ParticlesManager.java
package dclib.graphics;

import java.util.ArrayList;
import java.util.List;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Camera;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.ParticleEffect;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.math.Vector2;

import dclib.geometry.UnitConverter;

public final class ParticlesManager {

	private final TextureCache textureCache;
	private final Camera camera;
	private final Batch spriteBatch;
	private final UnitConverter unitConverter;
	// TODO: split this into another class
	private final List<ParticleEffect> particleEffects = new ArrayList<ParticleEffect>();

	public ParticlesManager(final TextureCache textureCache, final Camera camera, final Batch spriteBatch,
			final UnitConverter unitConverter) {
		this.textureCache = textureCache;
		this.camera = camera;
		this.spriteBatch = spriteBatch;
		this.unitConverter = unitConverter;
	}

	public final void createEffect(final String particleEffectPath, final Vector2 position) {
		ParticleEffect effect = new ParticleEffect();
		TextureAtlas atlas = textureCache.getAtlas("objects");
		effect.load(Gdx.files.internal("particles/" + particleEffectPath), atlas);
		effect.scaleEffect(unitConverter.getPixelsPerUnit());
		Vector2 pixelPosition = unitConverter.toPixelUnits(position);
		effect.setPosition(pixelPosition.x, pixelPosition.y);
		effect.start();
		particleEffects.add(effect);
	}

	public final void update(final float delta) {
		for (ParticleEffect particleEffect : new ArrayList<ParticleEffect>(particleEffects)) {
			particleEffect.update(delta);
			if (particleEffect.isComplete()) {
				particleEffects.remove(particleEffect);
			}
		}
	}

	public final void draw() {
		spriteBatch.setProjectionMatrix(camera.combined);
		spriteBatch.begin();
		for (ParticleEffect particleEffect : particleEffects) {
			particleEffect.draw(spriteBatch);
		}
		spriteBatch.end();
	}

}

File: dclib/core/src/main/java/dclib/physics/particles/ParticlesManager.kt
package dclib.physics.particles

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.ParticleEffect
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.Fixture
import com.badlogic.gdx.physics.box2d.World
import dclib.graphics.ScreenHelper
import dclib.graphics.TextureCache
import dclib.system.Updater
import java.util.*

class ParticlesManager(
        private val textureCache: TextureCache,
        private val spriteBatch: Batch,
        private val screenHelper: ScreenHelper,
        private val world: World
) : Updater {
    private val particleEffectDatas = ArrayList<ParticleEffectData>()

    fun createEffect(particleEffectPath: String, positionGetter: PositionGetter): ParticleEffect {
        val effect = ParticleEffect()
        val atlas = textureCache.getAtlas("objects")
        effect.load(Gdx.files.internal("particles/" + particleEffectPath), atlas)
        for (emitter in effect.emitters) {
            val box2dEmitter = ParticleEmitterBox2d(world, emitter, this::filterContact)
            effect.emitters.add(box2dEmitter)
            effect.emitters.removeValue(emitter, true)
        }
        effect.start()
        particleEffectDatas.add(ParticleEffectData(effect, positionGetter))
        return effect
    }

    override fun update(delta: Float) {
        for (particleEffectData in ArrayList(particleEffectDatas)) {
            val particleEffect = particleEffectData.effect
            val position = particleEffectData.positionGetter.get()
            particleEffect.setPosition(position.x, position.y)
            particleEffect.update(delta)
            if (particleEffect.isComplete) {
                particleEffectDatas.remove(particleEffectData)
            }
        }
    }

    fun draw() {
        screenHelper.setScaledProjectionMatrix(spriteBatch)
        spriteBatch.begin()
        for (particleEffectData in particleEffectDatas) {
            particleEffectData.effect.draw(spriteBatch)
        }
        spriteBatch.end()
    }

    private fun filterContact(fixture: Fixture): Boolean {
        return fixture.body.type == BodyDef.BodyType.StaticBody
    }

    private data class ParticleEffectData(
            val effect: ParticleEffect,
            val positionGetter: PositionGetter)
}
